### DISCUSSION TITLE
How to prevent a Resource from saving empty relationships

### DISCUSSION THREAD

**Discussion by griggsk** [filamentphp/filament#6440] [Help] [answered]
In the admin panel, I have users with optional demographics fields. For example, a user can select a level of education from a list or leave it blank. 

**Models**
```
User Model
     public function userEducation(): HasOne
    {
        return $this->hasOne(UserEducation::class);
    }

UserEducation Model
   public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
    public function education(): BelongsTo
    {
        return $this->belongsTo(Education::class);
    }

Education Model
  id|label

```
**My UserResource $form with the hasOne relationship field:**
 ```
Forms\Components\Section::make('Optional Demographics')
        ->relationship('userEducation')
        ->schema([
             Forms\Components\Select::make('education_id')
            ->options(Education::all()->pluck('label', 'id')) 
        ])
```
This all works great when the optional field is selected. The User and the UserEducation tables get populated correctly. 

But when left blank the UserEducation relationship tries to be created with a blank education_id and fails due to the foreign key requirement on the table. 

**Question: How do I prevent it from trying to save the relationship when it is blank?**

**Comment by danharrin** [comment, marked as answer]
Upgrade to `v2.17.37`. I've just added a new feature for this:

`->relationship('userEducation', fn (array $state): bool => filled($state['education_id'] ?? null))`

**Reply by griggsk** [reply]
Thanks, that worked perfectly on create. But on edit a user can not remove their selection.

**Reply by danharrin** [reply]
I don't think there is a good way for us to support that feature

**Reply by howdu** [reply]
@griggsk you could manually remove it from a [suffixAction](https://filamentphp.com/docs/2.x/forms/fields#affixes)

```php
Select::make('')
    ->suffixAction(fn ($record) => Action::make('delete-action')
        ->icon('heroicon-o-x')
        ->action(function () use ($record) {
           $record->userEducation()->delete();
        })
    ),
```

**Reply by danharrin** [reply]
Maybe we should do the same check when the record is saved, and if it fails then delete the record instead of not updating it

**Reply by griggsk** [reply]
A manual delete will suffice I think.  If I was doing this in straight PHP I would check on save and delete the record if the field changed from populated to null. 

I appreciated your responses and thanks for all your work!

**Reply by danharrin** [reply]
I've made it so that it deletes when the condition isn't satisfied in #6481
